File: integrations/jvm/eclipse/xomda-eclipse-plugin.test/src/org/xomda/eclipse/XomdaModelReaderTest.java
package org.xomda.eclipse;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;

class XomdaModelReaderTest {

    @Test
    void readsNameVersionAndPackages(@TempDir Path tmp) throws IOException {
        Path model = tmp.resolve("model.json");
        Files.writeString(model,
            "{\n" +
            "  \"name\": \"Demo\",\n" +
            "  \"version\": \"1.2.3\",\n" +
            "  \"packages\": [\n" +
            "    {\n" +
            "      \"name\": \"shop\",\n" +
            "      \"packages\": [\n" +
            "        { \"name\": \"billing\", \"entities\": [{\"name\": \"Invoice\"}] }\n" +
            "      ],\n" +
            "      \"entities\": [{\"name\": \"Product\"}],\n" +
            "      \"enums\": [{\"name\": \"Currency\"}]\n" +
            "    }\n" +
            "  ]\n" +
            "}\n"
        );

        XomdaModelReader.XomdaModel parsed = XomdaModelReader.read(model.toFile());
        assertEquals("Demo", parsed.name);
        assertEquals("1.2.3", parsed.version);
        assertEquals(1, parsed.packages.size());
        XomdaModelReader.XomdaPackage shop = parsed.packages.get(0);
        assertEquals("shop", shop.name);
        assertEquals(List.of("Product"), shop.entities.stream().map(e -> e.name).collect(Collectors.toList()));
        assertEquals(List.of("Currency"), shop.enums.stream().map(e -> e.name).collect(Collectors.toList()));
        assertEquals(1, shop.packages.size());
        assertEquals(List.of("Invoice"), shop.packages.get(0).entities.stream().map(e -> e.name).collect(Collectors.toList()));
    }

    @Test
    void toleratesMissingFields(@TempDir Path tmp) throws IOException {
        Path model = tmp.resolve("model.json");
        Files.writeString(model, "{}");
        XomdaModelReader.XomdaModel parsed = XomdaModelReader.read(model.toFile());
        assertEquals("(unnamed)", parsed.name);
        assertEquals(List.of(), parsed.packages);
    }
}

File: integrations/jvm/intellij/src/test/kotlin/org/xomda/intellij/XomdaModelReaderTest.kt
package org.xomda.intellij

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Files
import java.nio.file.Path

class XomdaModelReaderTest {

    @Test
    fun `reads name, version and packages`(@TempDir tmp: Path) {
        val model = tmp.resolve("model.json")
        Files.writeString(
            model,
            """
            {
              "name": "Demo",
              "version": "1.2.3",
              "packages": [
                {
                  "name": "shop",
                  "packages": [
                    { "name": "billing", "entities": [{"name": "Invoice"}] }
                  ],
                  "entities": [{"name": "Product"}],
                  "enums": [{"name": "Currency"}]
                }
              ]
            }
            """.trimIndent()
        )

        val parsed = XomdaModelReader.read(model.toFile())
        assertEquals("Demo", parsed.name)
        assertEquals("1.2.3", parsed.version)
        assertEquals(1, parsed.packages.size)
        val shop = parsed.packages.first()
        assertEquals("shop", shop.name)
        assertEquals(listOf("Product"), shop.entities.map { it.name })
        assertEquals(listOf("Currency"), shop.enums.map { it.name })
        assertEquals(1, shop.packages.size)
        assertEquals(listOf("Invoice"), shop.packages.first().entities.map { it.name })
    }

    @Test
    fun `tolerates missing fields`(@TempDir tmp: Path) {
        val model = tmp.resolve("model.json")
        Files.writeString(model, "{}")
        val parsed = XomdaModelReader.read(model.toFile())
        assertEquals("(unnamed)", parsed.name)
        assertEquals(emptyList<Any>(), parsed.packages)
    }
}
